### DISCUSSION TITLE
ATtiny167 SoftwareSerial  RX Pin

### DISCUSSION THREAD

**Discussion by Snolte1001** [SpenceKonde/ATTinyCore#674] [Support / Q & A]
Hello,
I use the Hardware UART of the ATtiny167 for LIN communication, this works without problems, but I also need the SoftwareSeial for communication with a Bluetooth chip.

If I define PA3 (INT1) as RX the SoftwareSerial works as expected, but I need this PIn for something else and therefore I want to use PB6 (INT0) as RX, but with this the serial reception of data does not work anymore although it is an interrupt pin...

Is this an bug or a feature, and is there a solution for it ?

**Comment by SpenceKonde** [comment]
Is this a Digispark Pro or a bare t167? The digispark uses PB3 and PB6 for USB, and that cicuitry may be interfering. 

SoftwareSerial does not use the INT pins. (would that it did). Intead, it uses the PCINTs (and claims every PCINT vector on the chip, which is one of the many reason I hate it).

**Comment by Snolte1001** [comment]
Hey, 
the Digispark Pro was my first try, there I removed the resistors to the USB socket as I don't need them, so the pin is free.
But also with a stand alone ATtiny167 it doesn't work either.

**Comment by SpenceKonde** [comment]
Weird. Well, it should work on *any* pin because all it needs is a PCINT, unless something is loading the pin..... I do not know why it would be failing on that pin in particular. Are you testing with the old release or the 2.0.0 public beta? There is no further work being done on the 1.5.x branch, all work is being done in 2.0.0-dev, and that will be merged into master when the small number of remaining issues are stamped out.

**Comment by Snolte1001** [comment]
That something else on my part loads the pin I can almost rule out,
I wrote a minimal program for testing that only waits for a character at the SoftwareSerial and returns it after a delay.
As I said, with PA3 it works, but not with PB6.

I will try another version tomorrow and then report.

**Comment by Snolte1001** [comment]
ok new findings, all pins of portA (0-7) work for RX ! but none of the pins of portB work

**Comment by SpenceKonde** [comment]
Wonderful! Thank you for that piece of information, I had actually typed up a question for you about taking an inventory of which pins do and do not work before I went to bed but then forgot to send it. 

It's generating the right ISRs.... (the assembly output is otherwise nigh incomprehensible because it's a class so it's always loading pointers to pointers - it's like trying to follow directions given as "Turn onto route 53, then take that to route 20", but the route numbers aren't on the street signs and you have no map. If there's a way to dump the map of how class ~properties~ member variables are located within the class objects, in ram, I don't know it (and neither does avr-objdump

Say, if you turn on the pullup on a pin on  PORTA (for example) PA3, but then ask for a  software serial port on the same pin of PORTB.... if you connect the incoming signal to PA3, does it receive anything? 

Can you do a manual install of ATTinyCore from the 2.0.0-dev branch, and see if the bug is still there? II would *much* rather investigate this on 2.0.0 sincethat;'s coming out in just a couple of weeks now, and it now has all the thing it had lacked for over a year such that it can be tested (it needed a completely new boards.txt and all new variant files, and the old version had two completely different cores depending on what part was selected, which needed to be merged into one, plus a bunch of problems ( https://github.com/SpenceKonde/ATTinyCore/milestone/16?closed=1 for an severely incomplete list and the changelog for a less-incomplete list: https://github.com/SpenceKonde/ATTinyCore/blob/v2.0.0-dev/ChangeLog.md )

Plus I'm not sure where my 167 board got off to, and all the remaining 167' I have are in QFN-20 package. and the breakout boards for those will be in the *next* order from the board house, so all I have are digispark pro's

**Comment by Snolte1001** [comment]
I think that if you ask a question to solve a problem, you should work independently on the solution, so I thought it is helpful to know on which pin problem occurs and which not ;-) 

I will try to install 2.0.0-dev, but I don't use the arduino IDE but VScode with platformio, where ATTinyCore 1.5.2 is a package of the Atmel AVR Platform.

I have already read that packages can be entered or overwritten in the platformio.ini, but I still have to find out how exactly this is done, do you have a tip? 

With digispark pro you should be able to troubleshoot, the package size doesn't matter...ok with digispark not all pins of PortB are free, but some are.

**Reply by SpenceKonde** [reply]
Oh damn, you're on platformio? 

Whelp, that's another thing that I need to make sure gets tested for 2.0.0 release isn't it? 
I do not believe 2.0.0-dev has been tested with platform IO at all. I know nothing about it other than that it can provide inputs that aren't possible with the Arduino IDE. 
There's a big difference between them that makes it really hard to provide a good UX on both: Arduino IDE you get defines passed to the compiler for board options via tools submenus. AFAIK, it's done with an ini on platformio, so that it is possible for combinations of defines to be present that are not possible to generate from the menus; the menus can enforce "exactly one of option X or option Y or option Z is defined" - but on platformio often none of them will be defined (because there's no documentation about what they're called, the documentation is for the Arduino IDE user). And the documentation is written for the user on the Arduino IDE, not platform IO, so there's no reference that includes the information for how to configure those options in platformio. But I don't know what needs to be in that documentation. Because I don't know how to use platform IO, you see.... and learning a new IDE well enough to document how to use my core with it is not something that I think I can pick up quickly... 

I'd like to have an understanding just so I can answer support inquiries - So many of my users are on vscode and/or platformio....  but there aren't enough hours in the day. 

Ugh.... This reminded me of the fact that I don't have CI working here because I don't know how to create a key portion of the yaml file and don't know where to turn to learn.  If I'm being honest, I basically have no idea whether most combinations of board + options will compile blink or even bare_minimum or if some of them error out due to typos or worse.. The ones I've worked on do, of course, but each chip probably has around 100 permutations of different options, and there are over 20 board options. I have no idea what I'm going to do for testing 2.0.0 frankly, Github actions even if I could get it working is not going to be able to cover enough test cases; I think I'm going to at minimum need to come up with a local test script that will use arduino-cli to build some test sketch or sketches that calls a bunch of key core functions, and run it overnight or something. I have a suspicion that the github action that gets used for CI is inefficient in the extreme (if it's not, then the overnight run is no good: it won't be done by the next morning.

**Comment by Snolte1001** [comment]
I have now installed 2.0.0-dev in Platformio, how to do this has been answered by the PlatformIO community:
https://community.platformio.org/t/use-another-version-of-the-attinycore/26301

but it didn't fix the problem, PortB still doesn't work for RX

**Comment by Snolte1001** [comment]
again new findings, with 2.0.0-dev RX on the hardware serial also no longer works, with 1.5.2 it works.

**Comment by SpenceKonde** [comment]
Whaaaa??? That is a very big surprise to me, as indidnt alter the LIN-as
usart code. Thank you for reporting this.

____________
Spence Konde
Azzy’S Electronics

New products! Check them out at tindie.com/stores/DrAzzy
GitHub: github.com/SpenceKonde
ATTinyCore: Arduino support for almost every ATTiny microcontroller
Contact: ***@***.***

**Comment by SpenceKonde** [comment]
Please direct further discussion to
https://github.com/SpenceKonde/ATTinyCore/issues/677
This sounds like a pair of criical bugs, bot of which on their own would be sufficient to block 2.0.0 release.

**Comment by SpenceKonde** [comment]
Software serial should be fixed in 2.0.0-dev, there was a problem with the macro to get the PCINT bit. 

Still unclear what's wrong with Serial though....

**Comment by SpenceKonde** [comment]
Definitely reprodiucing the problem with the lin port

**Comment by SpenceKonde** [comment]
Nevermind, I had the serial connector backwards. I cannot produce a problemwith the hardware serial port on 2.0.0-dev branch.
